feat: governance permissionless proposals (#260)
ethereum-optimism/design-docs: governance/permissionless-proposals.md

- The system remains non-plutocratic by relying on multiple delegate approvals instead of token thresholds alone.
- Additionally, certain roles (Security Council or a timelock) can still veto or cancel proposals if needed.

# Proposed Solution

At a high-level, we propose the creation of an `Top100DelegatesProposalValidator` contract that becomes the sole interface for proposal submissions to the Governor. The `Top100DelegatesProposalValidator` will:

- **Intercept proposal submissions**: Provide functions such as `propose()` and `proposeWithModule()` that include pre-validation steps.
- **Delegate approval verification**: Require that a proposal submission gathers the sufficient delegate sign-offs before moving it for voting. These approvals will be validated by setting a minimum threshold of voting power that a delegate needs in order to approve, this threshold will be set by the manager along with the voting window.
- **Submission window check:** Ensure the proposal is submitted within a defined short window relative to the current voting cycle. (Type proposal check, this is not valid of all types of proposals)

> - **Type-aware validation:** Confirm that the proposal meets any additional criteria based on its type (e.g. higher thresholds for larger OP grants) as defined in the `ProposalTypesConfigurator`.

Reviewer: Concretely, this is specified by the MVP tab [here](https://docs.google.com/spreadsheets/d/1HCAQQmBbWYpXni2zZ3Vnyv-bM_f5pHFFrpItI0TXdAk/edit?gid=1779711153#gid=1779711153)

Reviewer: I think the most significant change implied by the above is to add a `AttestationGatedProposal` which verifies proposer is allowed based on e.g. an attestation by OF

Author: Yes, we agree, for verifying an approved proposer we will use an attestation that is attested by OP Foundation that defines these addresses for each proposal type.

3. **Move to vote via Top100DelegatesProposalValidator:** With the final proposal ready, the top 100 delegates can start signing-off on the on-chain proposal using the `proposalId`. When the proposal is ready to move to voting anyone should be able to call `moveToVote` , and if the proposal meets the necessary rules, such as having the required top 100 delegates sign-offs, being submitted within the correct time window, not going over the distribution rate limit, and adhering to any proposal type-specific threshold it will be moved to the Governor for voting.
4. **Governor:** If the `Top100DelegatesProposalValidator`confirms that the proposal is valid, it forwards the proposal to the OP Governor contract. Then, the voting time begins as usual.

### Off-chain flow
```mermaid
  flowchart LR
    Delegate([Delegate])
    ForumDraft[OPGov Forum]
    Community([Community])
    ForumFinal[OPGov Forum]

    Delegate -->|submit DRAFT proposal| ForumDraft
    Community -->|provide feedback| ForumDraft

>     ForumDraft -->|submit FINAL proposal with proposalId| ForumFinal

Reviewer: for some reason the ` proposald` here is hidden in the markdown render

```

### On-chain flow
```mermaid
flowchart LR
    Proposer([Approved Proposer])
    Validator[ProposalValidator]
    TopDelegates([Top Delegates])
    Executor([Executioner])
    FinalValidator[ProposalValidator]
    OPGov[OPGov]

>     Proposer -->|propose| Validator

Reviewer: It seems like there is a tradeoff here between preventing a footgun where we validate certain things (e.g. OP amount) up front in this function call. However this does split validation logic between `propose` and `moveToVote`.  Maybe that's fine, feels worth discussing briefly

Reviewer: Another question is whether this two-step flow should exist for proposal types which do not require the two steps (e.g. Maintenance Upgrade Proposal can be validated/submitted immediately)

    Validator -->|get proposalId| Proposer
    TopDelegates -->|approve proposal| Validator

    Executor -->|move to vote| FinalValidator
    FinalValidator -->|propose| OPGov
```

# Alternatives Considered

- **Permissioned voting cycle configuration**: As an MVP, we allow a permissioned role to set the voting cycle and delegate approval thresholds. While not fully decentralized, this is acceptable given OF’s existing veto power. A future improvement could use an optimistic oracle to fetch these values on-chain, but this adds complexity we don’t believe is needed yet.

> - **Forum sign-offs with signature verification**: The current process relies on forum comments from top delegates as attestations. We considered maintaining this model, where proposers collect these off-chain signatures and verify them in the `Top100DelegatesProposalValidator`. However, direct on-chain approvals are simpler, reduce errors, and avoid coordination overhead.

Reviewer: Figuring out exactly how to submit these transactions is one open question we need to address.  Some options I can think of include:
- Chrome extension/bookmarklet to use within the forum
- Standalone local CLI or other tool which pulls data from a forum URL
- Standalone site which pulls data from forum URL
- Ability for proposer to signal proposal finality onchain somehow and allow others to then approve via etherscan or something

Notably we will need to support regular wallets and SAFEs here due to diversity of delegate wallets.

Author: Regarding the submission of the proposal, this can either happen through etherscan or Agora's UI.
For the approval of the proposal, by a top delegate, we suggest having this available through Agora's UI where they can display all the proposal data, that the delegate will be called to approve, by listening to an event that is being emitted when a proposal is successfully submitted.

# Risks and Uncertainties

> - **Proposal type flexibility**: The current design relies on a permissioned `ProposalTypesConfigurator` to define and update proposal types. This is acceptable for the MVP but introduces centralization risk. In the future, we’ll need a more permissionless or community-driven mechanism to support new proposal types without relying on a single role.

Reviewer 2: I believe this first MVP can have them separated and maybe merge them in a following iteration

# The Workflow

The general workflow would be:

1. **Draft Proposal:** A delegate starts by creating a Draft proposal and posting it on the optimism governance forum. At this stage, the proposal is open to the entire community for feedback. 
2. **Proposal Finalization**: After the proposal has been drafted and shared on the forum, delegates provide feedback, ask clarifying questions, and propose changes. Once the delegate incorporates this input and the proposal is refined, the delegate updates its status to "Final Draft". At this point, the delegate must create an on-chain transaction to create the proposal metadata record in the `ProposalValidator` by calling `submitProposal()`, which can be done either through Etherscan or Agora UI. This call emits a `ProposalSubmitted` event, which includes the `proposalId` and other relevant data. Agora listens for this event and uses it to fetch and display the finalized proposal on its UI, allowing delegates to begin the sign-off process.
    
```mermaid
flowchart LR

>     Delegate([Delegate]) -->|Submit DRAFT proposal| OPGov1[OPGov Forum]
>     Community([Community]) -->|Provide feedback| OPGov1
>     OPGov1 -->|Submit FINAL proposal| OPGov2[OPGov Forum]

Reviewer: ```suggestion
    Delegate([Delegate]) -->|Submit DRAFT proposal| OPGov[OPGov Forum]
    Community([Community]) -->|Provide feedback| OPGov
    OPGov -->|`submitProposal()` once FINAL| OnchainGov[ProposalValidator]
```

It should be this right?  I see the same diagram as current in markdown but I think just a typo, as the figma comment confirms "final == onchain"

# Purpose

This document outlines the design for enabling permissionless proposals in the Optimism governance system. The goal is to allow an approved address or anyone, based on the Proposal Type, to submit a proposal on-chain without requiring prior approval or gatekeeping by the manager. Instead, if all automated gating rules are met, the proposal is submitted for voting. 

# Summary

We propose introducing a `ProposalValidator` contract that sits in front of the existing Governor. This contract will enforce additional permissions rules such as verifying delegate approvals, checking submission windows, and validating the Proposal Type, before forwarding the proposal to the underlying governor. This design represents a step towards fully permissionless proposals. While it removes manual gatekeeping and automates rule enforcement on-chain, it continues to rely on configurable thresholds and permissions, set by a manager, to ensure system integrity. 

# Problem Statement + Context

> The current process, per the [**Operating Manual v0.4.2**](https://github.com/ethereum-optimism/OPerating-manual/blob/main/manual.md), involves proposals being drafted on the forum, receiving delegate approvals (explicit sign-offs from four of the top 100 delegates), and then being submitted for a vote during a specific voting cycle. The existing flow is partially off-chain and manually enforced: delegates post drafts in the forum, approvals are coordinated in the comments, and a designated manager is responsible for calling the Governor to formally start the voting cycle.

Reviewer 3: ```suggestion
The current process, per the [**Operating Manual v1.0.0**](https://github.com/ethereum-optimism/OPerating-manual/blob/main/manual.md), involves proposals being drafted on the forum, receiving delegate approvals (explicit sign-offs from four of the top 100 delegates), and then being submitted for a vote during a specific voting cycle. The existing flow is partially off-chain and manually enforced: delegates post drafts in the forum, approvals are coordinated in the comments, and a designated manager is responsible for calling the Governor to formally start the voting cycle.
```

| Governance Fund (Missions)    | Used to support Collective development and ecosystem growth via the Mission framework.                                                                                                                   | Permissionless                        | Approval                 | - Input valid proposal type<br>- Requested amount does not exceed the OP Cap                                         | - Gathered needed delegate approvals<br>- Voting cycle enforcement |
| Council Budget                | Allocation of funds for governance operations.                                                                                                                                                           | Permissionless                        | Approval                 | - Input valid proposal type<br>- Requested amount does not exceed the OP Cap                                         | - Gathered needed delegate approvals<br>- Voting cycle enforcement |

# The Workflow

> The general workflow would be:

Reviewer 3: The general workflow should use the language from the OPerating Manual (even if the OPerating Manual could also tighten its own use of language). I'll make more detailed comments below.

1. **Draft Proposal:** A delegate starts by creating a Draft proposal and posting it on the optimism governance forum. At this stage, the proposal is open to the entire community for feedback. 

> 2. **Proposal Finalization**: After the proposal has been drafted and shared on the forum, delegates provide feedback, ask clarifying questions, and propose changes. Once the delegate incorporates this input and the proposal is refined, the delegate updates its status to "Final Draft". At this point, the delegate must create an on-chain transaction to create the proposal metadata record in the `ProposalValidator` by calling `submitProposal()`, which can be done either through Etherscan or Agora UI. This call emits a `ProposalSubmitted` event, which includes the `proposalId` and other relevant data. Agora listens for this event and uses it to fetch and display the finalized proposal on its UI, allowing delegates to begin the sign-off process.

Reviewer 3: ```suggestion
2. **Proposal Finalization**: After the proposal has been drafted and shared on the forum, delegates provide feedback, ask clarifying questions, and propose changes. Once the delegate incorporates this input and the proposal is refined, the delegate updates its status to "Final". At this point, the delegate must create an on-chain transaction to create the proposal metadata record in the `ProposalValidator` by calling `submitProposal()`, which can be done either through Etherscan or Agora UI. This call emits a `ProposalSubmitted` event, which includes the `proposalId` and other relevant data. Agora listens for this event and uses it to fetch and display the finalized proposal on its UI, allowing delegates to begin the sign-off process.
```

> 2. **Proposal Finalization**: After the proposal has been drafted and shared on the forum, delegates provide feedback, ask clarifying questions, and propose changes. Once the delegate incorporates this input and the proposal is refined, the delegate updates its status to "Final Draft". At this point, the delegate must create an on-chain transaction to create the proposal metadata record in the `ProposalValidator` by calling `submitProposal()`, which can be done either through Etherscan or Agora UI. This call emits a `ProposalSubmitted` event, which includes the `proposalId` and other relevant data. Agora listens for this event and uses it to fetch and display the finalized proposal on its UI, allowing delegates to begin the sign-off process.

Reviewer 3: ```suggestion
2. **Proposal Finalization**: After the proposal has been drafted and shared on the forum, delegates provide feedback, ask clarifying questions, and propose changes. Once the delegate incorporates this input and the proposal is refined, the delegate updates its status to "Final Draft". At this point, the delegate must create an on-chain transaction to create the proposal metadata record in the `ProposalValidator` by calling `submitProposal()`, which can be done either through Etherscan or Agora UI. This call emits a `ProposalSubmitted` event, which includes the `proposalId` and other relevant data. Agora listens for this event and uses it to fetch and display the submitted proposal on its UI, allowing delegates to begin the sign-off process.
```

    
```mermaid
flowchart LR
    Delegate([Delegate]) -->|Submit DRAFT proposal| OPGov[OPGov Forum]
    Community([Community]) -->|Provide feedback| OPGov
    OPGov -->|submitProposal once FINAL| OnChainGov[ProposalValidator]
```
    
3. **On-chain Sign-Offs via Agora**: With the final proposal ready, the top 100 delegates can start signing-off on the on-chain proposal using the `proposalId` through Agora. Their sign-offs are recorded via the `approveProposal(...)` function in the `ProposalValidator` contract, that emits a `ProposalApproved` event. Agora will display the real-time signature status for each proposal.

> 4. **Move to vote via** `ProposalValidator`**:** Once the required number of sign-offs is reached, the "Approve Proposal" button in Agora changes to "Propose". Any user can click this button to call `moveToVote(...)` on the `ProposalValidator`. If the proposal meets the necessary rules, such as having the required top 100 delegates sign-offs, being submitted within the correct time window, not going over the distribution rate limit, and adhering to any proposal type-specific threshold it will be moved to the Governor for voting. This is done by internally calling `proposeWithModule(...)` **within `moveToVote(...)`.

Reviewer 3: ```suggestion
4. **Move to vote via** `ProposalValidator`**:** Once the required number of approvals is reached, the "Approve Proposal" button in Agora changes to "Propose". Any user can click this button to call `moveToVote(...)` on the `ProposalValidator`. If the proposal meets the necessary rules, such as having the required top 100 delegate approvals, being submitted within the correct time window, not going over the distribution rate limit, and adhering to any proposal type-specific threshold it will be moved to the Governor for voting. This is done by internally calling `proposeWithModule(...)` **within `moveToVote(...)`.
```

> 4. **Move to vote via** `ProposalValidator`**:** Once the required number of sign-offs is reached, the "Approve Proposal" button in Agora changes to "Propose". Any user can click this button to call `moveToVote(...)` on the `ProposalValidator`. If the proposal meets the necessary rules, such as having the required top 100 delegates sign-offs, being submitted within the correct time window, not going over the distribution rate limit, and adhering to any proposal type-specific threshold it will be moved to the Governor for voting. This is done by internally calling `proposeWithModule(...)` **within `moveToVote(...)`.

Reviewer 3: ```suggestion
4. **Move to vote via** `ProposalValidator`**:** Once the required number of sign-offs is reached, the "Approve Proposal" button in Agora changes to "Move To Vote". Any user can click this button to call `moveToVote(...)` on the `ProposalValidator`. If the proposal meets the necessary rules, such as having the required top 100 delegates sign-offs, being submitted within the correct time window, not going over the distribution rate limit, and adhering to any proposal type-specific threshold it will be moved to the Governor for voting. This is done by internally calling `proposeWithModule(...)` **within `moveToVote(...)`.
```

5. **Governor:** If the `ProposalValidator` checks that the proposal is valid, it forwards the proposal to the OP Governor contract. Then, the voting time begins as usual.

```mermaid
flowchart LR
    subgraph Agora UI
        AgoraUI -->|listens to ProposalSubmitted| A2[Agora UI]
    end

    A2 -->|Displays new proposal<br>based on Proposal Type| B[Agora UI]
    C[Top Delegates] -->|approve proposal Id| B
    B -->|approveProposal| D[ProposalValidator]

    E[Approved Proposer] -->|moveToVote| F[Agora UI]
    F -->|moveToVote| G[ProposalValidator]
    G -->|propose| H[OPGov]
```

> Detailed flows for each Proposal Type can be seen [here](https://www.figma.com/board/V0zh1Lg5QqhGfhVrJaIJVu/Permissionless-Proposals-DRAFT?node-id=0-1&p=f&t=sFuVhRioJR1kAIzV-0).

Reviewer 3: What about vetoes? How are they executed?

Reviewer 4: Well spotted! Addressed in this commit [7568bed](https://github.com/ethereum-optimism/design-docs/pull/260/commits/7568bed85f21511f188662b6699a7fbe47bdb457).

TL;DR: A veto is executed at the governor/timelock layer, not in `ProposalValidator`. If a veto is signalled or a security veto is required, the designated authority calls `OptimismGovernor.cancel(...)` before execution.

> # Alternatives Considered

Reviewer 3: Please fix language to match the OPerating Manual in the sections below:
"sign-off" -> "approve"
"Final Draft" -> "Final"
"propose(...)" -> "moveToVote(...)"

Each proposal type and their validation steps are detailed in the following table:

| **Proposal Type**             | **Description**                                                                                                                                                                                          | **Proposer**                          | **Proposal Voting Type** | **Submission Checks**                                                                                                | **Moving To Vote Checks**                                          |
|-------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------------------------------------|--------------------------|----------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------|

> | Protocol or Governor Upgrade  | Scheduled changes to the on-chain smart contracts comprising the mainnet Optimism protocol or governance contract.                                                                                       | Any approved address by OP Foundation | Optimistic                 | - Input valid proposal type<br>- Proposer is an approved address<br>- No funds are transferred<br>- No state changes | - Gathered needed delegate approvals                               |

Reviewer 3: What does "No state changes" mean here?

Protocol Upgrades (and Maintenance Upgrade Proposals) change the state in the L1 smart contracts, which is something we want to restrict (hence the permissioned model).

Governance Fund (Missions) and Council Budget do not change the state of the smart contracts, beyond moving funds around. It makes sense that those can be made permissionless with checks on the amount not exceeding the OP cap. These we should also require that don't cause any state changes in any contracts except the OP Token contract.

Reviewer 4: It means the proposal has an empty actions array and performs no on-chain calls that write storage or transfer assets. It’s signaling-only. Adding this to the design doc.

Reviewer 3: In that case the design docs must be wrong. We have `Protocol or Governor Upgrade` and `Maintenance Upgrade Proposals` as "No state changes", and those two definitely do on-chain calls that write to storage.

> | Maintenance Upgrade Proposals | Protocol maintenance required to continue network operation on a shorter timescale than voting periods. Must not materially change behavior for end users, infrastructure providers, or chain governors. | Any approved address by OP Foundation | Optimistic               | - Input valid proposal type<br>- Proposer is an approved address<br>- No funds are transferred<br>- No state changes | - N/A                                                              |

Reviewer 3: I guess that there isn't a difference really between Maintenance Upgrade Proposals and Protocol or Governor Upgrade, except that we don't want a vote on the former. There is then some trust that the approved addresses don't submit Maintenance Upgrade Proposals in bad faith.

Reviewer 4: Each attestation must specify the type of proposal that the user is allowed to submit (refer to the schema). To enable a user to submit both types of proposals (Protocol or Governance Upgrade / Maintenance Upgrade), two separate attestations are required. If a trusted proposer holds an attestation solely for Protocol or Governance types, they will not be permitted to submit a Maintenance Upgrade.

| Council Member Elections      | Elections to replace or renew signers in a given cohort.                                                                                                                                                 | Any approved address by OP Foundation | Ranked Choice            | - Input valid proposal type<br>- Proposer is an approved address<br>- No funds are transferred<br>- No state changes | - Gathered needed delegate approvals<br>- Voting cycle enforcement |
| Governance Fund (Missions)    | Used to support Collective development and ecosystem growth via the Mission framework.                                                                                                                   | Permissionless                        | Approval                 | - Input valid proposal type<br>- Requested amount does not exceed the OP Cap                                         | - Gathered needed delegate approvals<br>- Voting cycle enforcement |

> | Council Budget                | Allocation of funds for governance operations.                                                                                                                                                           | Permissionless                        | Approval                 | - Input valid proposal type<br>- Requested amount does not exceed the OP Cap                                         | - Gathered needed delegate approvals<br>- Voting cycle enforcement |

Reviewer 3: There is nothing to programmatically differentiate Governance Fund (Missions) from Council Budget types. Not really a problem, because delegates can not approve proposals using the wrong type, but I don't really know why we have two types here.

Reviewer 3: They might be useful for reporting reasons. I think that we can just add the observation that they are only conceptually different in the risks below.

Reviewer 3: Or maybe we don't need to say anything. It's not really a risk.

    
```mermaid
flowchart LR
    Delegate([Delegate]) -->|Submit DRAFT proposal| OPGov[OPGov Forum]
    Community([Community]) -->|Provide feedback| OPGov
    OPGov -->|submitProposal once FINAL| OnChainGov[ProposalValidator]
```
    
3. **On-chain approvals via Agora**: With the final proposal submitted, the top 100 delegates can start approving the on-chain proposal using the `proposalId` through Agora. Their approvals are recorded via the `approveProposal(...)` function in the `ProposalValidator` contract, that emits a `ProposalApproved` event. Agora will display the real-time signature status for each proposal.

> 4. **Move to vote via** `ProposalValidator`**:** Once the required number of approvals is reached, the "Approve Proposal" button in Agora changes to "Move To Vote". Any user can click this button to call `moveToVote(...)` on the `ProposalValidator`. If the proposal meets the necessary rules, such as having the required top 100 delegate approvals, being submitted within the correct time window, not going over the distribution rate limit, and adhering to any proposal type-specific threshold it will be moved to the Governor for voting. This is done by internally calling `proposeWithModule(...)` **within `moveToVote(...)`.

Reviewer 3: A bit of a UX nit, but the "Approve Proposal" should only change to "Move To Vote" if ALL the conditions pass, such as "having the required top 100 delegate approvals, being submitted within the correct time window, not going over the distribution rate limit, and adhering to any proposal type-specific threshold".

The button should stay in "Approve Proposal" while all of "being submitted within the correct time window, not going over the distribution rate limit, and adhering to any proposal type-specific threshold" are met, but "having the required top 100 delegate approvals" is not.

If any of "being submitted within the correct time window, not going over the distribution rate limit, and adhering to any proposal type-specific threshold" are not met, the "Approve Proposal" button should be disabled, and the reason shown.

Reviewer 4: > "Approve Proposal" should only change to "Move To Vote" if ALL the conditions pass
That is what is going to happen. We can clarify that in the step. Working on it.

- **Alligator voting power**: This design does not currently account for subdelegations via the Alligator contract. In a very rare case, a subdelegator could appear in the top 100 delegates by aggregating voting power from multiple authority chains. This introduces ambiguity, as subdelegated voting power may be partially consumed elsewhere.
- **Operational Monitoring and security responsibility**: As the `ProposalValidator` process becomes fully on-chain, it’s important to define which actors are responsible for monitoring its correct operation. For example, what conditions should be enforced or tracked by the smart contract itself, and which anomalies should be monitored off-chain by the Foundation or other ecosystem participants? If a bug or unexpected behavior occurs in `ProposalValidator`, it could block valid proposals from reaching the Governor or allow invalid ones through.

> - **Dynamic Attestation Service**: Since we heavily rely on the dynamic attestation service for updating the attestations of the top 100 delegates, it raises the concern of the service going offline and stopping the updating of attestations for a long period of time. Another concern is that currently the Schema used by the service uses a custom Resolver Contract that checks the issuer of the attestations, this should be managed only by the OP Foundation.

Reviewer 3: I would specify here the risk of the address approved to submit Maintenance Upgrade Proposals to do so to avoid a vote.
